woodwork/type_sys/inference_functions.py

```diff
@@ -74,5 +86,26 @@ def timedelta_func(series):
     return False
 
 
+def email_address_func(series: pd.Series) -> bool:
+    regex = ww.config.get_option('email_inference_regex')
+
+    # Includes a check for object dtypes
+    if not pdtypes.is_string_dtype(series.dtype):
+        return False
+
+    sample = get_inference_sample(series)
+    try:
+        sample_match_method = sample.str.match
+    except (AttributeError, TypeError):
+        # This can happen either when the inferred dtype for a series is not
+        # compatible with the pandas string API (AttributeError) *or* when the
+        # inferred dtype is not compatible with the string API `match` method
+        # (TypeError)
+        return False
+    matches = sample_match_method(pat=regex)
+
+    return matches.sum() == matches.count()
```

Comment:
Does this mean that all 10K row have to match the email regex?

Reply:
More specifically, it means that all valid (non-NA) rows have to match.  I had a brief discussion with @thehomebrewnerd about this.  He said we've tended to do an all or nothing type approach with the inference functions.  My interpretation of that was that all of the valid sampled records have to match the email address regex.  Originally, I had it so that only half of the sampled records needed to match and that percentage was configurable.  Happy to discuss it further if you have other ideas.

Reply:
Also, it appears that the behavior is slightly different depending on if the series came from a dask or koalas data set.  If it came from dask, then only 10k rows from the first partition in the dataset are considered (regardless of whether or not that partition actually has 10k rows).  If it came from koalas, then 10k rows are sampled from the result of a `.head(100000)` call.

Reply:
Actually, I had thought to mention the above difference to @thehomebrewnerd and ask if it seemed like that difference in behavior was okay.  Or is it a little confusing?  Not sure how I feel about it.  For what it's worth, I did have to structure some of the test series a specific way to account for the fact that the test series are split into two partitions when they're sent to dask.  So if you're trying to check if a bad record exists in some test series, it better be in the first partition of the series otherwise `TypeSystem.infer_logical_type` discards.